R/format_data.R

```diff
@@ -2414,6 +2414,676 @@ fmt_datetime <- function(data,
+
+  if (is.null(input_units)) {
+    return(NULL)
+  }
+
+  if (!is.character(input_units)) {
+
+    stop(
+      "The `input_units` input to `fmt_duration()` must be a character vector",
+      call. = FALSE
+    )
+  }
+
+  time_parts_vec <- c("weeks", "days", "hours", "mins", "minutes", "secs", "seconds")
+
+  if (!all(input_units %in% time_parts_vec) || length(input_units) != 1) {
+
+    stop(
+      "The value of `input_units` for `fmt_duration()` is invalid:\n",
+      "* Only one of `\"weeks\"`, `\"days\"`, `\"hours\"`, `\"minutes\"`, or \n",
+      "the `\"seconds\"` time part should be present",
+      call. = FALSE
+    )
+  }
+}
+
+normalize_duration_input_units <- function(input_units) {
+
+  # Ensure that key transforms occur
+  input_units %>%
+    tidy_sub("secs", "seconds") %>%
+    tidy_sub("mins", "minutes")
+}
+
+validate_duration_output_units <- function(output_units) {
+
+  if (!is.character(output_units)) {
+
+    stop(
+      "The `output_units` input to `fmt_duration()` must be a character vector",
+      call. = FALSE
+    )
+  }
+
+  if (length(output_units) < 1) {
+
+    stop(
+      "The `output_units` input to `fmt_duration()` must be a vector with at ",
+      "least one element",
+      call. = FALSE
+    )
+  }
+
+  time_parts_vec <- c("weeks", "days", "hours", "mins", "minutes", "secs", "seconds")
+
+  if (!all(output_units %in% time_parts_vec)) {
+
+    stop(
+      "There are invalid components in the `output_units` input to ",
+      "`fmt_duration()`:\n",
+      "* Only the `\"weeks\"`, `\"days\"`, `\"hours\"`, `\"minutes\"`, and\n",
+      "`\"seconds\"` time parts should be present",
+      call. = FALSE
+    )
+  }
+}
+
+normalize_duration_output_units <- function(output_units) {
+
+  # Ensure that key transforms occur and that the output units are a unique set
+  output_units <-
+    output_units %>%
+    tidy_sub("secs", "seconds") %>%
+    tidy_sub("mins", "minutes") %>%
+    unique()
+
+  # Ensure that the order of output units is from greatest to smallest
+  time_parts <- c("weeks", "days", "hours", "minutes", "seconds")
+  output_units[order(match(output_units, time_parts))]
+}
+
+values_to_durations <- function(
+    x,
+    in_units,
+    out_units,
+    out_style,
+    trim_zero_units,
+    max_output_units,
+    sep_mark,
+    dec_mark,
+    system,
+    locale,
+    patterns
+) {
+
+  # Obtain the units of `x` if it is of the difftime class
+  if (inherits(x, "difftime")) {
+    in_units <- units(x)
+  }
+
+  # Should `in_units` be anything other than days then
+  # convert all `x` values to days
+  if (in_units != "days") {
+
+    x <-
+      switch(
+        in_units,
+        weeks = x * 7,
+        hours = x / 24,
+        mins =,
+        minutes = x / 1440,
+        secs =,
+        seconds = x / 86400
+      )
+  }
+
+  x_str <- character(length(x))
+
+  for (i in seq_along(x)) {
+
+    x_val_i <- c()
+    x_str_i <- c()
+    x_rem_i <- abs(x[i])
```

Comment:
I think it would make sense to do a refactor of the variables here, so that we have:

1. A data frame that is initially `data.frame(value = numeric(0), time_part = character(0))`
2. `x_rem_i`, continues as is

The next for loop does nothing but build up that data frame (no formatting is performed).

Next, strip zero parts as appropriate.

Next, format (build x_str_i).

This has the advantages of making it harder for values and strings vectors from getting out-of-sync; and for colon-sep, we know the units for each resulting part.

Reply:
Now implemented.